### PR TITLE
Deal with undefined tensors in unbind backward

### LINKED ISSUE #9977: torch.stack() gradient errors in 0.4.1
## Issue description

We're seeing Pyro code breakage due to interaction of `torch.stack` and `binary_cross_entropy_with_logits` as used inside `torch.Bernoulli.log_prob()`.

Possibly related: https://github.com/pytorch/pytorch/issues/4274

## Code examples
The following examples were distilled from Pyro's [tests/contrib/tracking/test_em.py](https://github.com/uber/pyro/blob/pytorch-0.4.1-failing-tests/tests/contrib/tracking/test_em.py):
```py
x = torch.zeros(1, requires_grad=True)
y = torch.zeros(1)
stacked = torch.stack([y, x], -1)
loss = binary_cross_entropy_with_logits(stacked[:, 1], y, reduction='none')
g = torch.autograd.grad(loss, [x], create_graph=True)[0]
H = torch.autograd.grad(g.sum(), [x], create_graph=True)[0]
```
```
RuntimeError: dim() called on undefined Tensor
```
```py
x = torch.zeros(1, requires_grad=True)
y = torch.zeros(1)
stacked = torch.stack([x, y], -1)
loss = binary_cross_entropy_with_logits(stacked[:, 0], y, reduction='none')
g = torch.autograd.grad(loss, [x], create_graph=True)[0]
H = torch.autograd.grad(g.sum(), [x], create_graph=True)[0]
```
```
RuntimeError: Expected a Tensor of type Variable but found an undefined Tensor at position #1 for iterable argument #0 'tensors'
```
These are both equivalent to the following unstacked version which works fine:
```py
x = torch.zeros(1, requires_grad=True)
y = torch.zeros(1)
loss = binary_cross_entropy_with_logits(x, y, reduction='none')
g = torch.autograd.grad(loss, [x], create_graph=True)[0]
H = torch.autograd.grad(g.sum(), [x], create_graph=True)[0]
```

## System info
```
PyTorch version: 0.4.1
Is debug build: No
CUDA used to build PyTorch: None

OS: Mac OSX 10.13.3
GCC version: Could not collect
CMake version: version 3.9.6

Python version: 2.7
Is CUDA available: No
CUDA runtime version: No CUDA
GPU models and configuration: No CUDA
Nvidia driver version: No CUDA
cuDNN version: No CUDA

Versions of relevant libraries:
[pip] Could not collect
[conda] torch                     0.4.1                     <pip>
[conda] torchfile                 0.1.0                     <pip>
[conda] torchvision               0.2.1                     <pip>
```

### PR DESCRIPTION
**Pull request by t-vi**
When only part of the outputs of unbind are used in a backward,
the gradients for the others are undefined. This sets those
to zero in to_tensor_list.

Fixes: #9977

### CODE REVIEW

File: tools/autograd/templates/Functions.cpp [hunk 1]
```diff
@@ -51,7 +52,16 @@ void copy_range(variable_list& out, IndexRange range, at::ArrayRef<Tensor> t) {
 }
 
 std::vector<Tensor> to_tensor_list(const variable_list& variables) {
-  return fmap(variables, [](const Variable &v) { return static_cast<Tensor>(v); } );
+  IntList sizes;
```
Comment (SsnL) [inline tools/autograd/templates/Functions.cpp]: This doesn't match the name `to_tensor_list` anymore. :) Could you instead have a custom `unbind_backward`?

Comment (t-vi) [reply tools/autograd/templates/Functions.cpp]: Sure thing! Do you want to keep to_tensor_list?

Comment (SsnL) [reply tools/autograd/templates/Functions.cpp]: If it's not used elsewhere, feel free to delete it. :)

**Comment by SsnL** [comment]
The failure looks like that it might be related to the change. Can you see if it reproduces locally?

**Comment by t-vi** [comment]
Thanks! Seems to have worked. :)

**Review by ssnl — APPROVED** [review]

**Review by apaszke — APPROVED** [review]

**Review by soumith — APPROVED** [review]

**Pull request closed without merging**